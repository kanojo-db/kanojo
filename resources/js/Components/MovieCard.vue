<script lang="ts" setup>
import { Link } from '@inertiajs/inertia-vue3';
import { computed } from 'vue';
import { useI18n } from 'vue-i18n';

const props = defineProps({
    movie: {
        type: Object,
        required: true,
    },
});

const isVrMovie = computed(() => {
    return props.movie.type.name === 'VR' || props.movie.type_id === 4;
});

const posterUrl = computed(() => {
    return props.movie.media?.filter((m) => m.collection_name === 'poster')?.[0]
        ?.original_url;
});

const { t } = useI18n();

const averageScore = computed(() => {
    if (props.movie.love_reactant.reaction_total?.count) {
        const likeReactions = props.movie.love_reactant.reactions.filter(
            (reaction) => reaction.type.name === 'Like',
        );

        return (
            (likeReactions.length /
                props.movie.love_reactant.reaction_total.count) *
            100
        );
    }

    return 0;
});

const userScore = computed(() => {
    if (props.movie.love_reactant.reaction_total?.count) {
        return averageScore.value;
    }

    return t('web.general.not_rated');
});
</script>

<template>
    <Link
        class="block"
        :href="route('movies.show', { movie: movie.slug })"
        style="width: 200px"
    >
        <div class="fit column no-wrap justify-start items-start content-start">
            <q-img
                v-if="posterUrl"
                :src="posterUrl"
                width="200px"
                :ratio="2 / 3"
                :fit="isVrMovie ? 'contain' : 'cover'"
                class="rounded-borders bg-grey-1 shadow-1"
            />
            <div
                v-else
                class="row bg-grey-1 rounded-borders justify-center items-center shadow-1"
                style="width: 200px; height: 300px"
            >
                <q-icon
                    name="mdi-help"
                    size="150px"
                    color="grey-2"
                />
            </div>
            <div
                class="fit column no-wrap justify-start items-start content-start relative-position q-pt-sm"
            >
                <q-knob
                    :value="averageScore"
                    readonly
                    :min="0"
                    :max="10"
                    show-value
                    size="50px"
                    :thickness="0.15"
                    color="primary"
                    center-color="grey-2"
                    track-color="grey-2"
                    class="absolute"
                    style="top: -25px"
                >
                    <div
                        class="row justify-center items-start text-overline text-black"
                    >
                        <i18n-t
                            v-if="averageScore > 0"
                            keypath="web.general.x_percent"
                        >
                            <template #number>
                                <span class="text-weight-bolder">{{
                                    userScore
                                }}</span>
                            </template>
                        </i18n-t>
                        <span
                            v-else
                            class="text-weight-bolder text-h6"
                        >
                            {{ $t('web.general.not_rated') }}
                        </span>
                    </div>
                </q-knob>
                <span class="text-weight-bold q-mt-lg ellipsis-2-lines">{{
                    movie.title.en ? movie.title.en : movie.title.jp
                }}</span>
                <span class="q-mb-sm">{{ movie.product_code }}</span>
            </div>
        </div>
    </Link>
</template>
