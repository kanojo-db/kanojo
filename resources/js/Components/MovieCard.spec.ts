/**
 * @vitest-environment jsdom
 */
import type { Movie } from '@/types/models';

import messages from '@intlify/unplugin-vue-i18n/messages';
import { mount, renderToString } from '@vue/test-utils';
import { axe, toHaveNoViolations } from 'jest-axe';
import { describe, expect, it } from 'vitest';
import { createI18n } from 'vue-i18n';

import MovieCard from './MovieCard.vue';

expect.extend(toHaveNoViolations);

const movie: Movie = {
    id: 1,
    title: {
        'en-US': 'Movie Title',
        'ja-JP': '映画タイトル',
    },
    barcode: '1234567890123',
    release_date: '2021-01-01',
    length: 120,
    studio_id: 1,
    type_id: 1,
    slug: 'abc-123',
    created_at: '2021-01-01 00:00:00',
    updated_at: '2021-01-01 00:00:00',
    deleted_at: null,
    translations: {},
    reports: [],
    models: [],
    external_links: {},
    media: [],
    in_collection: false,
    in_favorites: false,
    in_wishlist: false,
    love_reactant_id: 1,
    love_reactant: {
        id: 1,
        reaction_counters: [],
        reactions: [],
        reaction_total: {
            count: 0,
            id: 1,
            weight: 1,
            reactant_id: 1,
            created_at: '2021-01-01 00:00:00',
            updated_at: '2021-01-01 00:00:00',
        },
        type: 'movie',
        created_at: '2021-01-01 00:00:00',
        updated_at: '2021-01-01 00:00:00',
    },
    studio: {
        id: 1,
        name: {
            'en-US': 'Studio Name',
            'ja-JP': 'スタジオ名',
        },
        founded_date: '2021-01-01',
        slug: 'studio-name',
        created_at: '2021-01-01 00:00:00',
        updated_at: '2021-01-01 00:00:00',
        deleted_at: null,
        translations: {},
        external_links: {},
        movies: {
            current_page: 1,
            data: [],
            first_page_url: '',
            from: 1,
            last_page: 1,
            last_page_url: '',
            links: [],
            next_page_url: null,
            path: '',
            per_page: 15,
            prev_page_url: null,
            to: 1,
            total: 1,
        },
        media: [],
    },
    type: {
        id: 1,
        name: 'Movie Type',
        created_at: '2021-01-01 00:00:00',
        updated_at: '2021-01-01 00:00:00',
        movies: [],
    },
    is_3d: false,
    is_vr: false,
    dmm_id: null,
    fanza_id: null,
    fc2_id: null,
    google_id: null,
    imdb_id: null,
    mgstage_id: null,
    tmdb_id: null,
    wikidata_id: null,
    fanart: null,
    poster: null,
    series: undefined,
};

const i18n = createI18n({
    legacy: false,
    locale: 'en-US',
    fallbackLocale: 'en-US',
    messages: messages,
});

describe('MovieCard', () => {
    it('renders properly in SSR mode', async () => {
        const contents = await renderToString(MovieCard, {
            global: {
                plugins: [i18n],
            },
            props: {
                movie,
            },
        });

        expect(contents).toMatchSnapshot();
    });

    it('has no a11y violations', async () => {
        const wrapper = mount(MovieCard, {
            global: {
                mocks: {
                    $route: vi.fn(),
                },
                plugins: [i18n],
            },
            props: {
                movie,
            },
        });

        const results = await axe(wrapper.html(), {
            rules: {
                region: { enabled: false },
            },
        });

        expect(results).toHaveNoViolations();
    });
});
