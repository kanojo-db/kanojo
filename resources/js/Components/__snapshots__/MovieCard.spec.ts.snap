// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MovieCard > renders properly in SSR mode 1`] = `"<a class=\\"relative block h-[250px] w-[100px] md:h-[435px] md:w-[200px]\\" href=\\"http://127.0.0.1/movies/abc-123\\"><div class=\\"flex flex-col flex-nowrap place-items-start\\"><div class=\\"flex aspect-[1/1.5] w-[100px] place-items-center justify-center rounded bg-stone-200 shadow-md md:w-[200px]\\"><svg viewBox=\\"0 0 24 24\\" width=\\"128\\" height=\\"128\\" class=\\"text-stone-300\\" color=\\"currentColor\\"><path fill=\\"currentColor\\" d=\\"M10 19h3v3h-3v-3m2-17c5.35.22 7.68 5.62 4.5 9.67c-.83 1-2.17 1.66-2.83 2.5C13 15 13 16 13 17h-3c0-1.67 0-3.08.67-4.08c.66-1 2-1.59 2.83-2.25C15.92 8.43 15.32 5.26 12 5a3 3 0 0 0-3 3H6a6 6 0 0 1 6-6Z\\"></path></svg></div><!--v-if--><div class=\\"relative flex flex-col flex-nowrap place-items-start pt-2 md:-top-10\\"><div class=\\"radial-progress text-primary bg-stone-200 font-black mr-2 hidden md:block\\" style=\\"--value:0;--size:4rem;--border-width:2px;\\" responsive><div class=\\"flex h-full items-center justify-center\\"><span class=\\"mb-1 text-stone-500\\"><!--[-->NR<!--]--></span></div></div><span class=\\"mt-4 text-sm md:text-base line-clamp-2 overflow-hidden overflow-ellipsis font-bold\\">Movie Title</span><span class=\\"mb-2\\">ABC-123</span></div></div></a>"`;
